File: lib/security-construct.ts
import { Construct } from 'constructs';
import { VpcConstruct } from './vpc-construct';
import { KmsKey } from '@cdktf/provider-aws/lib/kms-key';
import { IamRole } from '@cdktf/provider-aws/lib/iam-role';
import { IamRolePolicy } from '@cdktf/provider-aws/lib/iam-role-policy';

interface SecurityConstructProps {
  prefix: string;
  vpc: VpcConstruct;
}

export class SecurityConstruct extends Construct {
  public readonly kmsKeys: Record<string, KmsKey> = {};
  public readonly iamRoles: Record<string, IamRole> = {};
  public readonly securityGroups: Record<string, IamRole[]> = {};

  constructor(scope: Construct, id: string, props: SecurityConstructProps) {
    super(scope, id);

    // For each region, create KMS key, IAM roles with least privilege, security groups
    Object.keys(props.vpc.vpcs).forEach(region => {
      // KMS Key
      const kmsKeyInstance = new KmsKey(this, `${props.prefix}-kms-${region}`, {
        provider: props.vpc.providers[region],
        description: `KMS key for ${props.prefix} ${region} encryption`,
        keyUsage: 'ENCRYPT_DECRYPT',
        policy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Sid: 'Enable IAM User Permissions',
              Effect: 'Allow',
              Principal: { AWS: 'arn:aws:iam::*:root' },
              Action: 'kms:*',
              Resource: '*',
            },
          ],
        }),
        tags: {
          Name: `${props.prefix}-kms-key-${region}`,
          Environment: props.prefix,
        },
      });
      this.kmsKeys[region] = kmsKeyInstance;
      // IAM Role (EC2)
      const ec2RoleInstance = new IamRole(
        this,
        `${props.prefix}-ec2-role-${region}`,
        {
          provider: props.vpc.vpcs[region].provider,
          name: `${props.prefix}-ec2-role-${region}`,
          assumeRolePolicy: JSON.stringify({
            Version: '2012-10-17',
            Statement: [
              {
                Action: 'sts:AssumeRole',
                Effect: 'Allow',
                Principal: { Service: 'ec2.amazonaws.com' },
              },
            ],
          }),
          tags: {
            Name: `${props.prefix}-ec2-role-${region}`,
            Environment: props.prefix,
          },
        }
      );
      this.iamRoles[region] = ec2RoleInstance;
      // Attach least privilege policy to EC2
      new IamRolePolicy(this, `${props.prefix}-ec2-policy-${region}`, {
        provider: props.vpc.vpcs[region].provider,
        name: `${props.prefix}-ec2-policy-${region}`,
        role: ec2RoleInstance.id,
        policy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Action: [
                'logs:CreateLogGroup',
                'logs:CreateLogStream',
                'logs:PutLogEvents',
              ],
              Resource: `arn:aws:logs:${region}:*:*`,
            },
            {
              Effect: 'Allow',
              Action: ['kms:Decrypt', 'kms:DescribeKey'],
              Resource: kmsKeyInstance.arn,
            },
          ],
        }),
      });
      // IAM Role (Lambda)
      const lambdaRoleInstance = new IamRole(
        this,
        `${props.prefix}-lambda-role-${region}`,
        {
          provider: props.vpc.vpcs[region].provider,
          name: `${props.prefix}-lambda-role-${region}`,
          assumeRolePolicy: JSON.stringify({
            Version: '2012-10-17',
            Statement: [
              {
                Action: 'sts:AssumeRole',
                Effect: 'Allow',
                Principal: { Service: 'lambda.amazonaws.com' },
              },
            ],
          }),
          tags: {
            Name: `${props.prefix}-lambda-role-${region}`,
            Environment: props.prefix,
          },
        }
      );
      // Attach least privilege policy to Lambda
      new IamRolePolicy(this, `${props.prefix}-lambda-policy-${region}`, {
        provider: props.vpc.vpcs[region].provider,
        name: `${props.prefix}-lambda-policy-${region}`,
        role: lambdaRoleInstance.id,
        policy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Action: [
                'logs:CreateLogGroup',
                'logs:CreateLogStream',
                'logs:PutLogEvents',
              ],
              Resource: `arn:aws:logs:${region}:*:*`,
            },
            {
              Effect: 'Allow',
              Action: ['kms:Decrypt', 'kms:DescribeKey'],
              Resource: kmsKeyInstance.arn,
            },
            {
              Effect: 'Allow',
              Action: [
                'ec2:CreateNetworkInterface',
                'ec2:DescribeNetworkInterfaces',
                'ec2:DeleteNetworkInterface',
              ],
              Resource: '*',
            },
          ],
        }),
      });
    });
  }
}

File: lib/security_construct.py
"""security_construct.py
KMS encryption keys and security configurations.
"""

from constructs import Construct
import aws_cdk as cdk
from aws_cdk import (
    aws_kms as kms,
    aws_iam as iam
)


class SecurityConstruct(Construct):
    """
    Creates security resources including KMS keys and IAM policies
    for HIPAA-compliant healthcare data platform.
    """

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        environment_suffix: str,
        primary_region: str,
        dr_region: str,
        **kwargs
    ):
        super().__init__(scope, construct_id, **kwargs)

        # Create KMS key for encryption
        self.kms_key = kms.Key(
            self,
            f"EncryptionKey-{environment_suffix}",
            alias=f"healthcare-key-{environment_suffix}",
            description="KMS key for healthcare data encryption",
            enable_key_rotation=True,
            removal_policy=cdk.RemovalPolicy.DESTROY,
            pending_window=cdk.Duration.days(7)
        )

        # Grant necessary permissions for CloudWatch Logs
        self.kms_key.add_to_resource_policy(
            iam.PolicyStatement(
                sid="Allow CloudWatch Logs",
                principals=[
                    iam.ServicePrincipal(f"logs.{primary_region}.amazonaws.com")
                ],
                actions=[
                    "kms:Encrypt",
                    "kms:Decrypt",
                    "kms:ReEncrypt*",
                    "kms:GenerateDataKey*",
                    "kms:CreateGrant",
                    "kms:DescribeKey"
                ],
                resources=["*"],
                conditions={
                    "ArnLike": {
                        "kms:EncryptionContext:aws:logs:arn":
                        f"arn:aws:logs:{primary_region}:{cdk.Aws.ACCOUNT_ID}:*"
                    }
                }
            )
        )
